1.0.17 - Fallback when passed transition arguments on unsupported model

jackpot51 opened the pull request:
This will print an error message when unsupported firmware is selected, as opposed to simply reflashing the current firmware.

In the event of no transition being defined:
- If no argument is passed, it will update the current firmware variant.
- If --open is passed, it will require the system to be running System76 EC, or will print an error
- If --proprietary is passed, it will require the system to not be running System76 EC, or it will print an error

jacobgkau commented:
If I enter `sudo system76-firmware-cli --open schedule` (with the flag before the subcommand), I'm shown the message that `Only --open or --proprietary are supported`, which could be confusing since I did use --open.

ids1024 commented on src/transition.rs:
```diff
@@ -74,10 +74,29 @@ impl TransitionKind {
+        }
+
+        // Fallback in case transition is not defined
+        match self {
+            TransitionKind::Automatic => {
+                // Keep current firmware variant
+                Ok((model.to_string(), project.to_string()))
+            },
+            TransitionKind::Open => if project == "76ec" {
+                // Allow current firmware variant, it is already open
+                Ok((model.to_string(), project.to_string()))
+            } else {
+                Err(format!("Model '{}' is not supported by open firmware and EC at this time", model))
+            }
+            TransitionKind::Proprietary => if project != "76ec" {
+                // Allow current firmware variant, it is already proprietary
+                Ok((model.to_string(), project.to_string()))
+            } else {
+                Err(format!("Model '{}' is not supported by proprietary firmware and EC at this time", model))
```
It might be neater to condense this to something with less redundancy and nesting:

```rust
match self {
    TransitionKind::Open if project != "76ec" => {
        Err(format!("Model '{}' is not supported by open firmware and EC at this time", model))
    }
    TransitionKind::Proprietary if project == "76ec" => {
        Err(format!("Model '{}' is not supported by proprietary firmware and EC at this time", model))
    }
    _ => Ok((model.to_string(), project.to_string()))
}
```

But it should behave the same. Everything here looks sensible otherwise.

jackpot51 replied:
Yeah, that's cleaner, I will convert

jackpot51 commented:
Rebased on master, which includes the update to clap which should address @jacobgkau's comment. Also simplified the logic for the fallback, as suggested by @ids1024

jackpot51 commented:
Added a comment, "Make sure this is up to date with support article", to the list of supported models for transition
